### PR TITLE
Add opentelemetry tracer

### PR DESCRIPTION
**Pull request by k-yomo**
This PR implements opentelemetry tracer in reference to opencensus tracer.

### CODE REVIEW

**Comment by mdreizin** [comment]
@vektah @vvakame Could you please take a look at this PR?

File: gqlopentelemetry/tracer.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
+package gqlopentelemetry
+
+import (
+	"context"
+	"fmt"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/kv"
+	"google.golang.org/grpc/codes"
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:8]:
```suggestion
import (
	"context"
	"fmt"
	"go.opentelemetry.io/otel/api/global"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/label"
```

File: gqlopentelemetry/tracer.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			)
+		}
+	}
+
+	resp, err := next(ctx)
+
+	errList := graphql.GetFieldErrors(ctx, fc)
+	if len(errList) != 0 {
+		// UNKNOWN, HTTP Mapping: 500 Internal Server Error
+		span.SetStatus(codes.Unknown, errList.Error())
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:105]:
```suggestion
		span.SetStatus(codes.Error, errList.Error())
```

File: gqlopentelemetry/tracer.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	tracer := global.Tracer("")
+	ctx, span := tracer.Start(ctx, operationName(ctx))
+	defer span.End()
+	if !span.IsRecording() {
+		return next(ctx)
+	}
+
+	oc := graphql.GetOperationContext(ctx)
+	span.SetAttributes(
+		kv.String("request.query", oc.RawQuery),
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:43]:
```suggestion
		label.String("request.query", oc.RawQuery),
```

File: gqlopentelemetry/tracer.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	complexityStats, ok := oc.Stats.GetExtension(complexityExtension).(*extension.ComplexityStats)
+	if !ok {
+		// complexity extension is not used
+		complexityStats = &extension.ComplexityStats{}
+	}
+
+	if complexityStats.ComplexityLimit > 0 {
+		span.SetAttributes(
+			kv.Int64("request.complexityLimit", int64(complexityStats.ComplexityLimit)),
+			kv.Int64("request.operationComplexity", int64(complexityStats.Complexity)),
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:58]:
```suggestion
			label.Int64("request.complexityLimit", int64(complexityStats.ComplexityLimit)),
			label.Int64("request.operationComplexity", int64(complexityStats.Complexity)),
```

File: gqlopentelemetry/tracer.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+		)
+	}
+
+	for key, val := range oc.Variables {
+		span.SetAttributes(
+			kv.String(fmt.Sprintf("request.variables.%s", key), fmt.Sprintf("%+v", val)),
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:64]:
```suggestion
			label.String(fmt.Sprintf("request.variables.%s", key), fmt.Sprintf("%+v", val)),
```

File: gqlopentelemetry/tracer.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	defer span.End()
+	if !span.IsRecording() {
+		return next(ctx)
+	}
+
+	span.SetAttributes(
+		kv.String("resolver.path", fc.Path().String()),
+		kv.String("resolver.object", fc.Field.ObjectDefinition.Name),
+		kv.String("resolver.field", fc.Field.Name),
+		kv.String("resolver.alias", fc.Field.Alias),
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:84]:
```suggestion
		label.String("resolver.path", fc.Path().String()),
		label.String("resolver.object", fc.Field.ObjectDefinition.Name),
		label.String("resolver.field", fc.Field.Name),
		label.String("resolver.alias", fc.Field.Alias),
```

File: gqlopentelemetry/tracer.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	)
+	if a.DataDog {
+		span.SetAttributes(
+			// key from gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext#ResourceName
+			kv.String("resource.name", operationName(ctx)),
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:89]:
```suggestion
			label.String("resource.name", operationName(ctx)),
```

File: gqlopentelemetry/tracer.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		)
+	}
+	for _, arg := range fc.Field.Arguments {
+		if arg.Value != nil {
+			span.SetAttributes(
+				kv.String(fmt.Sprintf("resolver.args.%s", arg.Name), arg.Value.String()),
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:95]:
```suggestion
				label.String(fmt.Sprintf("resolver.args.%s", arg.Name), arg.Value.String()),
```

File: gqlopentelemetry/tracer.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	"go.opentelemetry.io/otel/api/kv"
+	"google.golang.org/grpc/codes"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/graphql/handler/extension"
+)
+
+type Tracer struct {
+	ComplexityExtensionName string
+	DataDog                 bool
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:16]:
```suggestion
	ComplexityExtensionName string
	DataDog                 bool
	Name                    string
```

File: gqlopentelemetry/tracer.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+func (a Tracer) ExtensionName() string {
+	return "OpenTelemetry"
+}
+
+func (a Tracer) Validate(schema graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	tracer := global.Tracer("")
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:34]:
```suggestion
	tracer := global.Tracer(a.Name)
```

**Review by mdreizin — CHANGES_REQUESTED** [review]
It would be nice to update `opentelemetry` to the latest version, update `go.mod`, `go.sum` and apply suggested changes.

Overall, it looks good to me. Great work! 👍🏻

File: gqlopentelemetry/tracer.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	"go.opentelemetry.io/otel/label"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/graphql/handler/extension"
+)
+
+type Tracer struct {
+	ComplexityExtensionName string
+	DataDog                 bool
+	Name                    string
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:17]: @k-yomo I was wrong about introducing `Name` field. I am sorry about that. I am going to test some changes locally and provide the feedback how we can use named tracer. Could you please delete `Name` for now?
```suggestion
```

File: gqlopentelemetry/tracer.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+func (a Tracer) ExtensionName() string {
+	return "OpenTelemetry"
+}
+
+func (a Tracer) Validate(schema graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	tracer := global.Tracer(a.Name)
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:35]:
```suggestion
	tracer := global.Tracer("")
```

**Comment by k-yomo** [comment]
@mdreizin Thank you for the review and the change suggestions. 
Updated open-telemetry client to the latest version.

**Review by mdreizin — CHANGES_REQUESTED** [review]

File: gqlopentelemetry/tracer.go (hunk 11 above)
Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: deleted!

Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: Looks like, tracer name is named by library not by user. 

https://github.com/open-telemetry/oteps/blob/master/text/0016-named-tracers.md#examples-of-tracer-and-meter-names

Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: Should it be `github.com/99designs/gqlgen-contrib`?

Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo Thanks for sharing!

Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo ie `const tracerName = "github.com/99designs/gqlgen-contrib/gqlopencensus"` and use it.

File: gqlopentelemetry/tracer.go [hunk 13]
```diff
@@ -31,7 +31,7 @@ func (a Tracer) Validate(schema graphql.ExecutableSchema) error {
 }
 
 func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
-	tracer := global.Tracer("")
+	tracer := global.Tracer("github.com/99designs/gqlgen-contrib")
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:34]: @k-yomo Please see https://github.com/99designs/gqlgen-contrib/pull/25#discussion_r510106125.

File: gqlopentelemetry/tracer.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+			label.String(fmt.Sprintf("request.variables.%s", key), fmt.Sprintf("%+v", val)),
+		)
+	}
+
+	return next(ctx)
+}
+
+func (a Tracer) InterceptField(ctx context.Context, next graphql.Resolver) (interface{}, error) {
+	fc := graphql.GetFieldContext(ctx)
+	tracer := global.Tracer("")
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:73]: @k-yomo Do you know if we need to set name here?

File: gqlopentelemetry/tracer.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+	"context"
+	"fmt"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/99designs/gqlgen/graphql/handler/extension"
+)
+
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:13]:
```suggestion

const tracerName = "github.com/99designs/gqlgen-contrib/gqlopentelemetry"

```

File: gqlopentelemetry/tracer.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+func (a Tracer) ExtensionName() string {
+	return "OpenTelemetry"
+}
+
+func (a Tracer) Validate(schema graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	tracer := global.Tracer("github.com/99designs/gqlgen-contrib")
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:34]:
```suggestion
	tracer := global.Tracer(tracerName)
```

File: gqlopentelemetry/tracer.go (hunk 14 above)
Comment (k-yomo) [reply gqlopentelemetry/tracer.go:73]: There is no other places to use the tracer name, so I think we don't need to make it const.

File: gqlopentelemetry/tracer.go (hunk 11 above)
Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: There are no other places to use the tracer name, so I think we don't need to make it const.

Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo Could you please set to `""` everywhere?

`opentelemetry` will use default name provided by `TracerProvider`:

```json
{
    "InstrumentationLibrary": {
        "Name": "go.opentelemetry.io/otel/sdk/tracer",
        "Version": ""
    }
}
```

File: gqlopentelemetry/tracer.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+func (a Tracer) ExtensionName() string {
+	return "OpenTelemetry"
+}
+
+func (a Tracer) Validate(schema graphql.ExecutableSchema) error {
+	return nil
+}
+
+func (a Tracer) InterceptResponse(ctx context.Context, next graphql.ResponseHandler) *graphql.Response {
+	tracer := global.Tracer("github.com/99designs/gqlgen-contrib/gqlopentelemetry")
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:34]:
```suggestion
	tracer := global.Tracer("")
```

**Review by mdreizin — CHANGES_REQUESTED** [review]

File: gqlopentelemetry/tracer.go (hunk 11 above)
Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: What's the reason to use default name?

Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo You can override it when you configure `TracerProvider`.

File: gqlopentelemetry/tracer.go (hunk 17 above)
Comment (mdreizin) [reply gqlopentelemetry/tracer.go:34]: @k-yomo Let's use empty string.

File: gqlopentelemetry/tracer.go (hunk 11 above)
Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: The user of this library can't override the name since there is  no way to manipulate.

Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: they set name in google's example.
https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/master/example/trace/http/client/client.go#L60

Setting name is not bad practice.

Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo That is true.

If you set it to `""` then `opentelemetry` will pick up the name from `TracerProvider`.

For now we can use `""` and refactor existing providers to use an approach where we can override `name`.

Do you agree?

Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: ` refactor existing providers to use an approach where we can override name`
Sorry, I don't get especially the above part. Do you mean, we would be better to provide a way to override provider's 'name' to user?

File: gqlopentelemetry/tracer.go (hunk 15 above)
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:13]:
```suggestion

const tracerName = "github.com/99designs/gqlgen-contrib/gqlopentelemetry"

```

File: gqlopentelemetry/tracer.go (hunk 17 above)
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:34]:
```suggestion
	tracer := global.Tracer(tracerName)
```

File: gqlopentelemetry/tracer.go [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+			label.String(fmt.Sprintf("request.variables.%s", key), fmt.Sprintf("%+v", val)),
+		)
+	}
+
+	return next(ctx)
+}
+
+func (a Tracer) InterceptField(ctx context.Context, next graphql.Resolver) (interface{}, error) {
+	fc := graphql.GetFieldContext(ctx)
+	tracer := global.Tracer("")
```
Comment (mdreizin) [inline gqlopentelemetry/tracer.go:73]:
```suggestion
	tracer := global.Tracer(tracerName)
```

**Review by mdreizin — CHANGES_REQUESTED** [review]

File: gqlopentelemetry/tracer.go (hunk 11 above)
Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo If we need to lock overriding tracer name, we also need to set it in two places. Please see my suggestions.

Comment (mdreizin) [reply gqlopentelemetry/tracer.go]: @k-yomo I agree with you. Please see my suggestions.

Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: `There are no other places to use the tracer name`
oh sorry I missed second one, and ↑ is incorrect.

Comment (k-yomo) [reply gqlopentelemetry/tracer.go]: Applied! Thank you for the suggestions!

**Comment by mdreizin** [comment]
@vektah @vvakame I am sorry for asking once again. Could you please to review this PR?

**Comment by mdreizin** [comment]
@k-yomo It looks like guys are busy and they can't chance to look at this PR right now.

Let's consider migrating it into https://github.com/open-telemetry/opentelemetry-go-contrib/tree/master/instrumentation/github.com

What do you think about migrating it into `opentelemetry-go-contrib`?

**Comment by mdreizin** [comment]
@k-yomo Let's do that. The migration plan is:

1. Add `go.mod` and `go.sum`
2. Adopt the code according ie https://github.com/open-telemetry/opentelemetry-go-contrib/blob/master/instrumentation/github.com/gin-gonic/gin/otelgin/gintrace.go
3. Add some tests if needed.

**Comment by mdreizin** [comment]
@k-yomo `go.opentelemetry.io` package has been recently updated (`v0.16.0`).

Could you please update your branch to use the latest version?

Here is migration guide:

1. Replace `go.opentelemetry.io/otel/api/global` with `go.opentelemetry.io/otel`
2. Replace `global.Tracer` with `otel.Tracer`

**Comment by ravilushqa** [comment]
does it hold?

**Comment by ravilushqa** [comment]
make [PR](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/761) to opentelemetry-go-contrib

**Comment by jay-manday** [comment]
Would really love to get this merged! this is awesome

**Comment by StevenACoffman** [comment]
I added the new home of this to the README. It's pretty great btw! Thanks for your work on this!

**Pull request closed without merging**